### ISSUE TITLE
[SUPPORT] Script to create local backups for the notes

### ISSUE THREAD

**Issue by ElfQrin** [qownnotes/scripts#202]
I'm writing a qml script to create local backups every time a note is updated and written on disk.

I'm aware that this code, should it work, would overwrite a backup note with a same name, because there aren't subdirectories in the backup directory, but this is just experimental, for now.

However, I have two questions:

- Is there a way to detect the system (Linux or Windows) and set the command "cp" or "copy" accordingly?
- Why script.startDetachedProcess in my code is not working (that is, the copy command apparently is not executed, and no error message is returned.)

`import QtQml 2.0

/*
 * QOwnNotes Local Backup
 * by Valerio Capello - labs.geody.com
 * This script will automatically back up a note locally every time a note is updated and written to the disk
 */
QtObject {
    property string copyCmd;
    property string bakNotesPath;
    property string fullNotePath;
    property variant settingsVariables: [
        {
            "identifier": "copyCmd",
            "name": "Copy Command (normally cp on Linux, copy on Windows)",
            "description": "",
            "type": "string",
            "default": "copy",
        },
        {
            "identifier": "bakNotesPath",
            "name": "Directory path to notes backup",
            "description": "",
            "type": "directory",
            "default": "C:/D/bak/QOwnNotes/bak_notes",
        }
    ];
    
    function onNoteStored(note) {
       bakNotesPath=script.toNativeDirSeparators(bakNotesPath + "/");
       fullNotePath=script.toNativeDirSeparators(note.fullNoteFilePath);
       
       script.log("Copying " + "\"" + fullNotePath + "\"" + " to " + "\"" + bakNotesPath + note.fileName + ".bak" + "\"");
       script.startDetachedProcess(copyCmd , "\"" + fullNotePath + "\"" + " " + "\"" + bakNotesPath + note.fileName + ".bak"+ "\"");
    }
}
`

[auto_bak_updated_notes.zip](https://github.com/qownnotes/scripts/files/12794510/auto_bak_updated_notes.zip)

**Comment by pbek** [comment]
> Is there a way to detect the system (Linux or Windows) and set the command "cp" or "copy" accordingly?

https://www.qownnotes.org/scripting/methods-and-objects.html#check-whether-platform-is-linux-os-x-or-windows

> script.startDetachedProcess(copyCmd , "\"" + fullNotePath + "\"" + " " + "\"" + bakNotesPath + note.fileName + ".bak"+ "\"");

You are not passing the parameters correctly, please see: https://www.qownnotes.org/scripting/methods-and-objects.html#starting-an-external-program-in-the-background

**Comment by ElfQrin** [comment]
I've read all the examples, and I can't make the copy command work with startDetachedProcess.

I've tried to pass parameters as an array of strings (comma separated):
` script.startDetachedProcess(copyCmd , ["\"" + fullNotePath + "\"" , "\"" + bakNotesPath + note.fileName + ".bak" + "\""] , "callback-cp" , 1);`

and as a single string (that is, passing both parameters as a single parameter separated by a space character, as in the cp/copy command):
` script.startDetachedProcess(copyCmd , ["\"" + fullNotePath + "\"" + " " + "\"" + bakNotesPath + note.fileName + ".bak" + "\""] , "callback-cp" , 1);`

I'm also using the callback to check if I can get anything wrong from there, but it looks fine there.

Code:
`
import QtQml 2.0

/*
 * QOwnNotes Local Backup r2023-10-03 fr2023-10-03
 * by Valerio Capello - labs.geody.com
 * This script will automatically back up a note locally every time a note is updated and written to the disk
 * Known issues:
 * - Since there are no subfolders in the back up directory, if the name of the backed up notes clashes with the name of another note in a different folder, the backup note (not the original ones) will be overwritten.
 */

QtObject {
    property string copyCmd;
    property string bakNotesPath;
    property string fullNotePath;
    property variant settingsVariables: [
        {
            "identifier": "bakNotesPath",
            "name": "Directory path to notes backup",
            "description": "",
            "type": "directory",
            "default": "C:/D/bak/QOwnNotes/bak_notes",
        }
    ];
    
    function onNoteStored(note) {
       if (script.platformIsWindows()) {
           copyCmd="copy";
       } else {
           copyCmd="cp";
       }
    
       bakNotesPath=script.toNativeDirSeparators(bakNotesPath + "/");
       fullNotePath=script.toNativeDirSeparators(note.fullNoteFilePath);
       // fullBakNotesPath = bakNotesPath + note.fileName + ".bak";
       
       script.log("Backing up " + "(" + copyCmd + ") " + "\"" + fullNotePath + "\"" + " to " + "\"" + bakNotesPath + note.fileName + ".bak" + "\"");
       script.startDetachedProcess(copyCmd , ["\"" + fullNotePath + "\"" , "\"" + bakNotesPath + note.fileName + ".bak" + "\""] , "callback-cp" , 1);
    }
    
    function onDetachedProcessCallback(callbackIdentifier, resultSet, cmd, thread) {
         if (callbackIdentifier == "callback-cp") {
             script.log(`${cmd[0]} ${cmd[1]}`);
         }
    }
 
}
`

[auto_bak_updated_notes.zip](https://github.com/qownnotes/scripts/files/12796057/auto_bak_updated_notes.zip)

**Comment by ElfQrin** [comment]
By the way, is there any built in copy funcion in qml? Like File.copy(source, dest)?

**Comment by pbek** [comment]
Best look how others have done it: https://github.com/search?q=repo%3Aqownnotes%2Fscripts%20script.startDetachedProcess&type=code

> By the way, is there any built in file copy funcion in qml? Like File.copy(source, dest)?

No, there isn't. 😉

**Comment by ElfQrin** [comment]
> Best look how others have done it: https://github.com/search?q=repo%3Aqownnotes%2Fscripts%20script.startDetachedProcess&type=code

I already looked, but to no avail because the format of the command I used was correct.

I found it more useful to write a script that would write to a file the arguments it had received, so that I realized that qml automatically adds quotes to an argument if it contains spaces in it.

Another problem is that the "copy" command on Windows (I haven't tried "cp" on Linux yet) does not seem to want to accept the arguments passed by qml, if instead with the same function I call up a batch file that in turn passes the same arguments to copy, everything works as expected.

cp.bat
`
@ echo off
copy %1 %2 %3
`
[auto_bak_updated_notes.zip](https://github.com/qownnotes/scripts/files/12803340/auto_bak_updated_notes.zip)
[cp.zip](https://github.com/qownnotes/scripts/files/12803341/cp.zip)
